### ISSUE TITLE
Hazelcast Java client not resolving member address when trying to reconnect

### ISSUE THREAD

**Issue by ycordier-pro** [hazelcast/hazelcast#17062]
Context: basic HZ setup in a docker environment : 1 server + 1 client, HZ configuration based on tcp-ip with network aliases (using docker service names that are resolved against docker internal DNS).
When HZ server restarts, docker reassigns a new IP address to the container but the client forever tries to reconnect to the old IP address without re-resolving the member name to obtain the new IP.

**Version used :** 3.12.7 (also tested on 4.0.1 : same issue)
**Cluster size :** single member
**Number of clients :** 1 Java client
**Java version :** 1.8 on client side / 11 on server side (Note : DNS network ttl cache is set to 0 everywhere)
**OS :** Host: CentOS with docker engine (Docker version 19.03.6). Client container with debian based image. Server container from public HZ image: Alpine Linux v3.8.
**Logs :** see below in the steps

**Steps to reproduce :**
Using Docker (docker-compose) :
Create a HZ single member cluster using the standard public image, with tcp-ip member configured with the Docker service alias names (https://hub.docker.com/r/hazelcast/hazelcast/)
Server configuration example :
```
            <tcp-ip enabled="true">
              <member-list>
                <member>hazelcast:5701</member>
              </member-list>
            </tcp-ip>
```
("hazelcast" here is the service name in docker-compose for the HZ server, it's just a standard network alias that gets resolved on the docker network)

Connect a Java client configured with cluster-members filled with the same Docker service alias name.
Client configuration example :
```
    <network>
        <cluster-members>
            <address>hazelcast:5701</address>
        </cluster-members>
    </network>
```
On server startup, HZ successfully starts up. Logs shows the member configured as:
```
hazelcast     | Members {size:1, ver:1} [
hazelcast     |      Member [hazelcast]:5701 - ed12db96-27b4-4f05-b99b-840799a6bebf this
hazelcast     | ]
```
On client startup, HZ resolves the IP address of the server and connects to the cluster:
```
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to cluster with name: hazelcast
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.internal-3] com.hazelcast.client.connection.ClientConnectionManager.null hz.client_1 [hazelcast] [3.12.7] Setting ClientConnection{alive=true, connectionId=1, channel=NioChannel{/172.20.0.40:57827->hazelcast/172.20.0.39:5701}, remoteEndpoint=[hazelcast]:5701, lastReadTime=2020-06-08 12:25:44.343, lastWriteTime=2020-06-08 12:25:44.317, closedTime=never, connected server version=3.12.7} as owner with principal ClientPrincipal{uuid='cd0aefaa-df16-4699-88b5-a16e9f476c5d', ownerUuid='239fa3d4-8e13-4f3f-a11e-f3539db2a6fb'}
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.internal-3] com.hazelcast.client.connection.ClientConnectionManager.null hz.client_1 [hazelcast] [3.12.7] Authenticated with server [hazelcast]:5701, server version:3.12.7 Local address: /172.20.0.40:57827
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.event-5] com.hazelcast.client.spi.impl.ClientMembershipListener.null hz.client_1 [hazelcast] [3.12.7]
hz-test-client              |
hz-test-client              | Members [1] {
hz-test-client              |      Member [hazelcast]:5701 - 239fa3d4-8e13-4f3f-a11e-f3539db2a6fb
hz-test-client              | }
hz-test-client              |
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.cluster-] com.hazelcast.core.LifecycleService.null hz.client_1 [hazelcast] [3.12.7] HazelcastClient 3.12.7 (20200430 - bf4b643) is CLIENT_CONNECTED
```
Everything works perfectly until you restart the HZ server service.
Note that on Docker, with the start/stop of a container, the IP address changes.
The client then looses the connection to the server and tries to reconnect.
But at this point the client will try to reconnect keeping the old IP address it found at startup time (although the JVM dns cache is set to 0).
In the logs, while the server is being restarted, I can see :
```
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.hz.client_1.IO.thread-in-0] com.hazelcast.client.connection.nio.ClientConnection.null hz.client_1 [hazelcast] [3.12.7] ClientConnection{alive=false, connectionId=1, channel=NioChannel{/172.20.0.40:57827->hazelcast/172.20.0.39:5701}, remoteEndpoint=[hazelcast]:5701, lastReadTime=2020-06-08 12:25:59.755, lastWriteTime=2020-06-08 12:25:54.447, closedTime=2020-06-08 12:25:59.755, connected server version=3.12.7} closed. Reason: Connection closed by the other side
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.hz.client_1.IO.thread-in-0] com.hazelcast.client.connection.ClientConnectionManager.null hz.client_1 [hazelcast] [3.12.7] Removed connection to endpoint: [hazelcast]:5701, connection: ClientConnection{alive=false, connectionId=1, channel=NioChannel{/172.20.0.40:57827->hazelcast/172.20.0.39:5701}, remoteEndpoint=[hazelcast]:5701, lastReadTime=2020-06-08 12:25:59.755, lastWriteTime=2020-06-08 12:25:54.447, closedTime=2020-06-08 12:25:59.755, connected server version=3.12.7}
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.client_1.cluster-] com.hazelcast.core.LifecycleService.null hz.client_1 [hazelcast] [3.12.7] HazelcastClient 3.12.7 (20200430 - bf4b643) is CLIENT_DISCONNECTED
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to cluster with name: hazelcast
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:25:59 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: Connection refused to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:25:59 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 1 of 2147483647.
hz-test-client              | 2020-06-08 12:26:05 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:10 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketTimeoutException
hz-test-client              | 2020-06-08 12:26:10 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 2 of 2147483647.
hz-test-client              | 2020-06-08 12:26:15 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:20 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketTimeoutException
hz-test-client              | 2020-06-08 12:26:20 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 3 of 2147483647.
hz-test-client              | 2020-06-08 12:26:25 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:30 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketTimeoutException
hz-test-client              | 2020-06-08 12:26:30 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 4 of 2147483647.
hz-test-client              | 2020-06-08 12:26:35 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:40 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketTimeoutException
hz-test-client              | 2020-06-08 12:26:40 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 5 of 2147483647.
hz-test-client              | 2020-06-08 12:26:45 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:49 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: No route to host to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:26:49 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 6 of 2147483647.
hz-test-client              | 2020-06-08 12:26:54 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:57 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: No route to host to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:26:57 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 7 of 2147483647.
hz-test-client              | 2020-06-08 12:27:03 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:27:06 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: No route to host to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:27:06 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 8 of 2147483647.
hz-test-client              | 2020-06-08 12:27:11 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:27:14 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: No route to host to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:27:14 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 9 of 2147483647.
```
but 172.20.0.39 is the old IP address of the hazelcast server, it has changed in docker since the restart and the client never attempts to resolve "hazelcast" again, leading to impossible reconnection.
It seems that when the client is executing its reconnection mechanism, it reuses the same connection object with the IP address resolved at startup time, but if it has changed since, it will never attempt to resolve the member address defined in the configuration again to reconnect.

I also tried the same thing with a 2 member cluster and it leads to the same errors.
On the server side whenever a member IP address changes, other members are capable of reconnecting to it automatically and see the address change, the issue is only present on the Java client side.
Tried with/without smart-routing but doesn't affect client behavior on this.
Tried with "hazelcast.diagnostics.enabled" on client side but didn't see anything more that would point me to a configuration issue or something onto which I can do anything.

Is this the expected behavior? 
Thanks for the feedback!

**Comment by pertsodian** [comment]
As per https://groups.google.com/forum/#!msg/hazelcast/6OM_msgJvzg/kDi06y29CQAJ I think it is intended. I continue to use 3.11.x until I figure out how to work with it.

**Comment by sancar** [comment]
@ycordier-pro @pertsodian We will be rediscussing what to do. My answer to this here still applies, https://groups.google.com/forum/#!msg/hazelcast/6OM_msgJvzg/kDi06y29CQAJ 

The only thing that comes to my mind is that give a property for the users that wants the old behavior. I don't actually want to do that because that behavior is not tested and supported. Giving an option will be like officially supporting the use case, which actually we did not spend time on maturing. 

We will discuss this within the team and let you know here.

**Comment by pertsodian** [comment]
Is allowing cluster failover feature with single cluster only for the non-enterprise version an option?

If yes, we can potentially configure limited reconnect attempts (that fits the pod health check reasonably), then clear all caches and perform cluster failover. But since only one cluster is configured, the client will to connect to same cluster again, but resolve fresh IP address this time.

Does that sound ok? Real value of cluster failover is disaster recovery via multiple clusters is still an enterprise feature.

**Comment by sancar** [comment]
@pertsodian The problem is that from the client perspective there is no difference between a restarted cluster or two clusters. We have reiterated this with the team and for now, we don't want to give a buggy feature to open source users. (so we don't want to restore the behaviour to pre 3.12 ) 
This use-case is supported by our Enterprise Blue/Green Feature.

**Comment by pertsodian** [comment]
Just for my understanding, why is "from the client perspective there is no difference between a restarted cluster or two clusters" a problem?

**Comment by sancar** [comment]
@pertsodian If it was possible to differentiate, I could of implement your request on open source version for a single cluster use case. But I can not. 
In more detail, In Blue/Green, you can start a second cluster together with the first one. Possibly migrate the data first via WAN replication. Then close the first one/ blacklist the clients in the first cluster via the Management Center. 
`Close the first one` case is the same as restart a single cluster from the client perspective.

**Comment by ycordier-pro** [comment]
Thanks all for your comments and the google group link.
In fact the client is handled within the HZ Tomcat session manager in my case, so I believe there is some state change listener to add somewhere in https://github.com/hazelcast/hazelcast-tomcat-sessionmanager so that it's able to restart a new client if the first one shuts down, to allow for cluster IP refresh.
I'll clone the repo and maybe propose a patch, this shouldn't be too difficult if not yet implemented.

**Comment by ycordier-pro** [comment]
After some more tests with a 2 members cluster, I think there is still an issue especially in the docker context.
Let's consider the following cluster configuration :
```
            <tcp-ip enabled="true">
              <member-list>
                <member>hazelcast1:5701</member>
                <member>hazelcast2:5701</member>
              </member-list>
            </tcp-ip>
```
and on the client side :
```
    <network>
        <cluster-members>
            <address>hazelcast1:5701</address>
            <address>hazelcast2:5701</address>
        </cluster-members>
    </network>
```
At startup, the HZ client will connect to the first member in the list, hazelcast1. Everything works fine.
Now let's say hazelcast1 member fails and is being restarted (with a new IP since it's a container), the client will have switched to hazelcast2 member which is ok at this point. But let's say you shut down hazelcast2 member (or it fails for some other reason), then the client will NEVER be able to connect back to hazelcast1 because it will keep connecting to the previous IP address it had in his cache for this member, even if in the hazelcast2 server logs I can see that in the meantime it successfully detected the member reconnection with the new IP address, it seems that the fact that hazelcast1 has changed its IP doesn't get propagated to the clients which seem to be only relying on the name to match it to what they have in connection cache.

In such a scenario, with the state listener, one could restart the whole client but IMHO this should be taken care of by the client automatically as it's just a simple failover case (no blue/green or whatsoever).

What do you think ?

**Comment by sancar** [comment]
This use case indeed deserves a fix. I will try to look into this issue in more detail next week.

**Comment by burakcelebi** [comment]
Hi, we have discovered that there are significant consequences in these use cases. We need more time to come to a decision. Thanks for your patience :)

**Comment by ycordier-pro** [comment]
Sure, no hurry, take the necessary time ;)

**Comment by ycordier-pro** [comment]
Hi, just a FYI, after [implementing the proposed solution on the tomcat replication session HZ manager](https://github.com/hazelcast/hazelcast-tomcat-sessionmanager/pull/92) (which is to listen for client shutdown and recreate a new one when necessary), I'm still experiencing some weird behaviors where client thinks it's still connected to a member whereas the member has already restarted and changed its IP address.
I must say this seems to happen only in race conditions where I restart the members multiple times in random ways and orders (I have a 2 member cluster in my test case).

In such case I'm getting errors like this one:
```
hz-test-client              | com.hazelcast.core.OperationTimeoutException: ClientInvocation{clientMessage = ClientMessage{connection=null, length=25597, operation=Map.Set, isRetryable=false, correlationId=1359, messageType=10f00, isEvent=false, isfragmented=false}, objectName = hz-test-client, target = partition 88, sendConnection = null} timed out because exception occurred after client invocation timeout 120000 ms. Current time: 2020-07-15 18:23:03.565. Start time: 2020-07-15 18:21:03.407. Total elapsed time: 120158 ms.
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.notifyException(ClientInvocation.java:264)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.invokeOnSelection(ClientInvocation.java:173)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.retry(ClientInvocation.java:193)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.run(ClientInvocation.java:179)
hz-test-client              |      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
hz-test-client              |      at java.util.concurrent.FutureTask.run(FutureTask.java:266)
hz-test-client              |      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
hz-test-client              |      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
hz-test-client              |      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
hz-test-client              |      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
hz-test-client              |      at java.lang.Thread.run(Thread.java:748)
hz-test-client              |      at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
hz-test-client              |      at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
hz-test-client              | Caused by: java.io.IOException: No available connection to member 63b91b70-6eaf-4d8b-9833-83a6550822d5
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.SmartClientInvocationService.getConnection(SmartClientInvocationService.java:125)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.SmartClientInvocationService.invokeOnTarget(SmartClientInvocationService.java:118)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.SmartClientInvocationService.invokeOnPartitionOwner(SmartClientInvocationService.java:95)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.invokeOnSelection(ClientInvocation.java:166)
hz-test-client              |      ... 11 more
```

In this case the client doesn't seem to somehow "mark" the connection as bad/lost (so that it falls in shutdown state, which would be then detected by the code I've added and would be restarted), the only way I'm able to get it back to life is to either restart the client application (the tomcat server) or restart the full HZ cluster (restart all the members, 2 actually in my test case).

Otherwise the solution seem to fix 99% of the issue and the tomcat replication plugin is now able to automatically reconnect whenever cluster members gets restarted, there seem to be only some few rare cases where it still falls in the above case. I'm trying to get a 100% reproducable scenario meanwhile.

**Comment by sancar** [comment]
Hi all, we have decided to apply the fix but since it requires more testing to make sure it is stable, it is postponed to the next patch(4.1.1)
